daemon/session_agent.go

```diff
@@ -0,0 +1,88 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+type SessionAgent struct {
+	Version  string
+	listener net.Listener
+	serve    *shutdownServer
+	tomb     tomb.Tomb
+	router   *mux.Router
+
+	mu sync.Mutex
+}
+
+func (s *SessionAgent) Init() error {
+	listenerMap, err := activationListeners()
+	if err != nil {
+		return err
+	}
+	agentSocket := fmt.Sprintf("%s/%d/snap-session.socket", dirs.XdgRuntimeDirBase, os.Getuid())
+	if listener, err := getListener(agentSocket, listenerMap); err == nil {
+		s.listener = &ucrednetListener{listener}
```

Comment:
why wrap it in `ucrednet`? are you using that anywhere (it doesn't look like you are).

Reply:
I obviously haven't implemented it here, but I was thinking to add some code to double check who we're talking to.

If everything is set up correctly, this shouldn't be necessary: permissions on `$XDG_RUNTIME_DIR` should prevent other users from connecting to the socket, and AppArmor rules should prevent confined apps connecting.  However, it is probably worth double checking in case those protections have failed.

With that said, I agree that `ucrednet` is probably the wrong tool here.  We don't want request level checking: we'd just want to drop unexpected connections immediately.  That'd probably be simpler and more reliable to implement independently, yes.